aiida_quantumespresso/calculations/__init__.py

```diff
@@ -379,12 +377,26 @@ def _generate_PWCPinputdata(self, parameters, settings, pseudos, structure, kpoi
                 fixed_coords_strings.append(
                     "  {:d} {:d} {:d}".format(*if_pos_values))
 
-        for site, fixed_coords_string in zip(
-                structure.sites, fixed_coords_strings):
-            atomic_positions_card_list.append(
-                "{0} {1:18.10f} {2:18.10f} {3:18.10f} {4}\n".format(
-                    site.kind_name.ljust(6), site.position[0], site.position[1],
-                    site.position[2], fixed_coords_string))
+        if use_fractional:
+            import numpy as np
+            atomic_positions_card_list = ["ATOMIC_POSITIONS crystal\n"]
+            cell = np.array(structure.cell)
+            abs_pos = np.array([_.position for _ in structure.sites])
+            rel_pos = np.dot(abs_pos, np.linalg.inv(cell))
+            for site, site_rel_coords, fixed_coords_string in zip(
```

Comment:
It looks like this for-loop can be abstracted as it is pretty much identical in both cases.
The only thing that you need to put inside the conditional is the `atomic_positions_card_list` and then a variable `coordinates` which contains either fractional or absolute.

Reply:
Agreed, I moved the for loop outside the conditional.